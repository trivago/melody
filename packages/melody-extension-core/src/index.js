/**
 * Copyright 2017 trivago N.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS-IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { unaryOperators, binaryOperators, tests } from './operators';
import { AutoescapeParser } from './parser/autoescape';
import { BlockParser } from './parser/block';
import { DoParser } from './parser/do';
import { EmbedParser } from './parser/embed';
import { ExtendsParser } from './parser/extends';
import { FilterParser } from './parser/filter';
import { FlushParser } from './parser/flush';
import { ForParser } from './parser/for';
import { FromParser } from './parser/from';
import { IfParser } from './parser/if';
import { ImportParser } from './parser/import';
import { IncludeParser } from './parser/include';
import { MacroParser } from './parser/macro';
import { SetParser } from './parser/set';
import { SpacelessParser } from './parser/spaceless';
import { UseParser } from './parser/use';
import { MountParser } from './parser/mount';

import forVisitor from './visitors/for';
import testVisitor from './visitors/tests';
import filters from './visitors/filters';
import functions from './visitors/functions';

const filterMap = [
    'attrs',
    'classes',
    'styles',
    'batch',
    'escape',
    'format',
    'merge',
    'nl2br',
    'number_format',
    'raw',
    'replace',
    'reverse',
    'round',
    'striptags',
    'title',
    'url_encode',
    'trim',
].reduce((map, filterName) => {
    map[filterName] = 'melody-runtime';
    return map;
}, Object.create(null));

Object.assign(filterMap, filters);

const functionMap = [
    'attribute',
    'constant',
    'cycle',
    'date',
    'max',
    'min',
    'random',
    'range',
    'source',
    'template_from_string',
].reduce((map, functionName) => {
    map[functionName] = 'melody-runtime';
    return map;
}, Object.create(null));
Object.assign(functionMap, functions);

export const extension = {
    tags: [
        AutoescapeParser,
        BlockParser,
        DoParser,
        EmbedParser,
        ExtendsParser,
        FilterParser,
        FlushParser,
        ForParser,
        FromParser,
        IfParser,
        ImportParser,
        IncludeParser,
        MacroParser,
        SetParser,
        SpacelessParser,
        UseParser,
        MountParser,
    ],
    unaryOperators,
    binaryOperators,
    tests,
    visitors: [forVisitor, testVisitor],
    filterMap,
    functionMap,
};

export {
    AutoescapeBlock,
    BlockStatement,
    BlockCallExpression,
    MountStatement,
    DoStatement,
    EmbedStatement,
    ExtendsStatement,
    FilterBlockStatement,
    FlushStatement,
    ForStatement,
    ImportDeclaration,
    FromStatement,
    IfStatement,
    IncludeStatement,
    MacroDeclarationStatement,
    VariableDeclarationStatement,
    SetStatement,
    SpacelessBlock,
    AliasExpression,
    UseStatement,
    UnaryNotExpression,
    UnaryNeqExpression,
    UnaryPosExpression,
    BinaryOrExpression,
    BinaryAndExpression,
    BitwiseOrExpression,
    BitwiseXorExpression,
    BitwiseAndExpression,
    BinaryEqualsExpression,
    BinaryNotEqualsExpression,
    BinaryLessThanExpression,
    BinaryGreaterThanExpression,
    BinaryLessThanOrEqualExpression,
    BinaryGreaterThanOrEqualExpression,
    BinaryNotInExpression,
    BinaryInExpression,
    BinaryMatchesExpression,
    BinaryStartsWithExpression,
    BinaryEndsWithExpression,
    BinaryRangeExpression,
    BinaryAddExpression,
    BinaryMulExpression,
    BinaryDivExpression,
    BinaryFloorDivExpression,
    BinaryModExpression,
    BinaryPowerExpression,
    BinaryNullCoalesceExpression,
    TestEvenExpression,
    TestOddExpression,
    TestDefinedExpression,
    TestSameAsExpression,
    TestNullExpression,
    TestDivisibleByExpression,
    TestConstantExpression,
    TestEmptyExpression,
    TestIterableExpression,
} from './types';
