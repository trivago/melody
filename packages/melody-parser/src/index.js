/**
 * Copyright 2017 trivago N.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS-IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import Parser from './Parser';
import TokenStream from './TokenStream';
import * as Types from './TokenTypes';
import Lexer from './Lexer';
import { EOF, CharStream } from './CharStream';
import { LEFT, RIGHT } from './Associativity';
import {
    setStartFromToken,
    setEndFromToken,
    copyStart,
    copyEnd,
    copyLoc,
    getNodeSource,
    createNode,
    hasTagStartTokenTrimLeft,
    hasTagEndTokenTrimRight,
    isMelodyExtension,
} from './util';

function parse(code, options, ...extensions) {
    return createExtendedParser(code, options, ...extensions).parse();
}

function createExtendedParser(code, options, ...extensions) {
    let passedOptions = options;
    const passedExtensions = extensions;
    if (isMelodyExtension(options)) {
        // Variant without options parameter: createExtendedParser(code, ...extensions)
        passedOptions = undefined;
        passedExtensions.unshift(options);
    }
    const lexer = createExtendedLexer(code, options, ...passedExtensions);
    const parser = new Parser(
        new TokenStream(lexer, passedOptions),
        passedOptions
    );
    for (const ext of passedExtensions) {
        parser.applyExtension(ext);
    }
    return parser;
}

function createExtendedLexer(code, options, ...extensions) {
    let passedOptions = options;
    const passedExtensions = extensions;
    if (isMelodyExtension(options)) {
        // Variant without options parameter: createExtendedLexer(code, ...extensions)
        passedOptions = undefined;
        passedExtensions.unshift(options);
    }
    const lexer = new Lexer(new CharStream(code), passedOptions);
    for (const ext of passedExtensions) {
        lexer.applyExtension(ext);
    }
    return lexer;
}

export {
    Parser,
    TokenStream,
    Lexer,
    EOF,
    CharStream,
    LEFT,
    RIGHT,
    parse,
    createExtendedLexer,
    createExtendedParser,
    setStartFromToken,
    setEndFromToken,
    copyStart,
    copyEnd,
    copyLoc,
    getNodeSource,
    createNode,
    hasTagStartTokenTrimLeft,
    hasTagEndTokenTrimRight,
    Types,
};
