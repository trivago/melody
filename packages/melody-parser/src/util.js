/**
 * Copyright 2017 trivago N.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS-IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
export function setStartFromToken(node, { pos: { index, line, column } }) {
    node.loc.start = { line, column, index };
    return node;
}

export function setEndFromToken(node, { pos: { line, column }, end }) {
    node.loc.end = { line, column, index: end };
    return node;
}

export function setMarkFromToken(
    node,
    propertyName,
    { pos: { index, line, column } }
) {
    node[propertyName] = { line, column, index };
    return node;
}

export function copyStart(
    node,
    {
        loc: {
            start: { line, column, index },
        },
    }
) {
    node.loc.start.line = line;
    node.loc.start.column = column;
    node.loc.start.index = index;
    return node;
}

export function copyEnd(node, end) {
    node.loc.end.line = end.loc.end.line;
    node.loc.end.column = end.loc.end.column;
    node.loc.end.index = end.loc.end.index;
    return node;
}

export function getNodeSource(node, entireSource) {
    if (entireSource && node.loc.start && node.loc.end) {
        return entireSource.substring(node.loc.start.index, node.loc.end.index);
    }
    return '';
}

export function copyLoc(node, { loc: { start, end } }) {
    node.loc.start.line = start.line;
    node.loc.start.column = start.column;
    node.loc.start.index = start.index;
    node.loc.end.line = end.line;
    node.loc.end.column = end.column;
    node.loc.end.index = end.index;
    return node;
}

export function createNode(Type, token, ...args) {
    return setEndFromToken(setStartFromToken(new Type(...args), token), token);
}

export function startNode(Type, token, ...args) {
    return setStartFromToken(new Type(...args), token);
}

export function hasTagStartTokenTrimLeft(token) {
    return token.text.endsWith('-');
}

export function hasTagEndTokenTrimRight(token) {
    return token.text.startsWith('-');
}

export function isMelodyExtension(obj) {
    return (
        obj &&
        (Array.isArray(obj.binaryOperators) ||
            typeof obj.filterMap === 'object' ||
            typeof obj.functionMap === 'object' ||
            Array.isArray(obj.tags) ||
            Array.isArray(obj.tests) ||
            Array.isArray(obj.unaryOperators) ||
            Array.isArray(obj.visitors))
    );
}
