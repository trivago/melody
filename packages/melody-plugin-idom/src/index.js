/**
 * Copyright 2017 trivago N.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS-IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import * as t from 'babel-types';
import mountVisitor from './visitors/mount.js';
import idomVisitor from './visitors/idom.js';

export default {
    visitors: [idomVisitor, mountVisitor],
    filterMap: {
        raw(path) {
            if (path.parentPath.is('PrintStatement')) {
                path.replaceWithJS(
                    t.expressionStatement(
                        t.callExpression(
                            t.identifier(
                                path.state.addImportFrom('melody-idom', 'raw')
                            ),
                            [path.get('target').node]
                        )
                    )
                );
            } else {
                path.replaceWithJS(
                    t.callExpression(
                        t.identifier(
                            path.state.addImportFrom('melody-idom', 'rawString')
                        ),
                        [path.get('target').node]
                    )
                );
            }
        },
    },
};
