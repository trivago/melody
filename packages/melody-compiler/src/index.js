/**
 * Copyright 2017 trivago N.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS-IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { CharStream, Lexer, TokenStream, Parser } from 'melody-parser';
import { traverse, merge, Path, Scope } from 'melody-traverse';

import { Template, File } from './Template.js';
import analyse from './analyse/index.js';
import convert from './convert/index.js';
import State from './state/State.js';
import finalizer from './finalizer/index.js';

import * as t from 'babel-types';
import { CodeGenerator } from 'babel-generator';

// workaround for https://github.com/rollup/rollup/issues/430
import { TokenTypes } from 'melody-parser';
export { Template, File, TokenTypes };

export function compile(fileName: String, source: String, ...extensions) {
    const root = parseString(fileName, source, ...extensions);
    const template = new Template(root.expressions),
        file = new File(fileName, template),
        state = new State(file, source);
    let analyseVisitor = analyse,
        convertVisitor = convert;
    for (const ext of extensions) {
        if (ext.visitors) {
            if (Array.isArray(ext.visitors)) {
                for (const visitor of (ext.visitors: Array)) {
                    if (visitor.analyse) {
                        analyseVisitor = merge(analyseVisitor, visitor.analyse);
                    }
                    if (visitor.convert) {
                        convertVisitor = merge(convertVisitor, visitor.convert);
                    }
                }
            } else {
                if (ext.visitors.analyse) {
                    analyseVisitor = merge(
                        analyseVisitor,
                        ext.visitors.analyse
                    );
                }
                if (ext.visitors.convert) {
                    convertVisitor = merge(
                        convertVisitor,
                        ext.visitors.convert
                    );
                }
            }
        }
        if (ext.filterMap) {
            Object.assign(state.filterMap, ext.filterMap);
        }
        if (ext.functionMap) {
            Object.assign(state.functionMap, ext.functionMap);
        }
        if (ext.options) {
            Object.assign(state.options, ext.options);
        }
    }
    convertVisitor = merge(convertVisitor, finalizer);
    const scope = Scope.get(
        Path.get({
            container: file,
            key: 'template',
        })
    );
    traverse(file, analyseVisitor, scope, state);
    traverse(file, convertVisitor, scope, state);
    return t.file(file.template);
}

function parseString(fileName: string, source: string, ...extensions) {
    const lexer = new Lexer(new CharStream(source));
    for (const ext of extensions) {
        lexer.applyExtension(ext);
    }
    const parser = new Parser(new TokenStream(lexer));
    for (const ext of extensions) {
        parser.applyExtension(ext);
    }

    return parser.parse();
}

export function toString(jsAst, code) {
    const g = new CodeGenerator(jsAst, {}, code);
    // Babel sanitises strings to not contain complex characters
    // however we need them in order to be able to render complex strings
    g._generator.StringLiteral = function(node, parent) {
        var raw = this.getPossibleRaw(node);
        if (!this.format.minified && raw != null) {
            this.token(raw);
            return;
        }

        return this.token(JSON.stringify(node.value));
    };
    return g.generate();
}
