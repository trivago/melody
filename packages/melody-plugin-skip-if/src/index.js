/**
 * Copyright 2017 trivago N.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS-IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import * as t from 'babel-types';
import { Node, type, alias, visitor } from 'melody-types';
import template from 'babel-template';
import { Types, setStartFromToken, setEndFromToken } from 'melody-parser';

export class SkipIfBlock extends Node {
    constructor(target, expressions) {
        super();
        this.target = target;
        this.expressions = expressions;
    }
}
type(SkipIfBlock, 'SkipIfBlock');
alias(SkipIfBlock, 'Block');
visitor(SkipIfBlock, 'expressions');

export const SkipIfParser = {
    name: 'skip',
    parse(parser, token) {
        const tokens = parser.tokens;

        tokens.expect(Types.SYMBOL, 'if');
        const condition = tokens.expect(Types.SYMBOL);
        tokens.expect(Types.TAG_END);

        const skipIfBlock = new SkipIfBlock(
            condition,
            parser.parse((tokenText, token, tokens) => {
                return !!(
                    token.type === Types.TAG_START &&
                    tokens.nextIf(Types.SYMBOL, 'endskip')
                );
            }).expressions
        );

        setStartFromToken(skipIfBlock, token);
        setEndFromToken(skipIfBlock, tokens.expect(Types.TAG_END));

        return skipIfBlock;
    },
};

const buildSkipIfDefined = template(`
if (customElements.get(ELEMENT_NAME) !== undefined) {
    SKIP
} else {
    BODY
}
`);

export default {
    tags: [SkipIfParser],
    visitors: [
        {
            analyse: {
                SkipIfBlock(path) {
                    if (path.node.target.text === 'client') {
                        path.skip();
                    }
                },
            },
            convert: {
                SkipIfBlock: {
                    enter(path) {
                        if (path.node.target.text === 'client') {
                            path.replaceWithJS(skip(this));
                        }
                    },
                    exit(path) {
                        const target = path.node.target.text;
                        if (target === 'server') {
                            path.replaceWithMultipleJS(
                                ...path.node.expressions
                            );
                        } else if (target === 'defined') {
                            const el = path.findParentPathOfType('Element');
                            const elementName = el.node.name;
                            if (elementName.indexOf('-') === -1) {
                                this.error(
                                    'skip if defined can only be used inside of a custom element',
                                    el.node.elementNameLoc,
                                    `Custom Element must contain a "-" within their name but "${elementName}" does not seem to contain one.
More information about custom elements can be found here: https://developer.mozilla.org/en-US/docs/Web/Web_Components/Custom_Elements`,
                                    elementName.length
                                );
                            }
                            path.replaceWithJS(
                                buildSkipIfDefined({
                                    ELEMENT_NAME: t.stringLiteral(elementName),
                                    SKIP: skip(this),
                                    BODY: path.node.expressions,
                                })
                            );
                        } else {
                            this.error(
                                'Unknown skip if target',
                                path.node.target.pos,
                                `The target of the skip must be either 'client', 'server' or 'defined' but was "${target}" instead.`,
                                target.length
                            );
                        }
                    },
                },
            },
        },
    ],
};

function skip(state) {
    return t.expressionStatement(
        t.callExpression(
            t.identifier(state.addImportFrom('melody-idom', 'skip')),
            []
        )
    );
}
