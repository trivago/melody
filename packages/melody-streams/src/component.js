/**
 * Copyright 2018 trivago N.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS-IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { enqueueComponent } from 'melody-idom';
import { shallowEqual } from './util/shallowEqual';
import { combine } from './operators/combine';
import { BehaviorSubject, Subject, timer } from 'rxjs';
import {
    distinctUntilChanged,
    tap,
    take,
    catchError,
    ignoreElements,
} from 'rxjs/operators';
import { of } from 'rxjs';

const warningTimer = timer(500).pipe(
    tap(() => {
        /* eslint-disable no-console */
        console.warn(
            'Warning: Your Component did not emit any state updates for at least 500ms.'
        );
        /* eslint-enable no-console */
    }),
    take(1),
    ignoreElements()
);

function Component(element) {
    // part of the public API
    this.el = element;
    this.refs = {};
    this.props = {};
    this.propsStream = new BehaviorSubject({});
    this.updates = new Subject();
    this.subscriptions = [];
    this.state = {};
}

Object.assign(Component.prototype, {
    type: 'streaming',
    apply(props) {
        this.propsStream.next(props);
        if (this.subscriptions.length === 0) {
            const t = this.getTransform({
                dispatchCustomEvent(eventName, detail, options = {}) {
                    const event = new CustomEvent(eventName, {
                        ...options,
                        detail,
                    });
                    this.el.dispatchEvent(event);
                },
                props: this.propsStream,
                updates: this.updates,
                subscribe: obs => this.subscriptions.push(obs.subscribe()),
            });
            const warningSubscription = warningTimer.subscribe();
            const s = t.pipe(distinctUntilChanged(shallowEqual)).subscribe(
                state => {
                    if (!warningSubscription.closed)
                        warningSubscription.unsubscribe();
                    this.state = state;
                    enqueueComponent(this);
                },
                err => {
                    if (process.env.NODE_ENV !== 'production') {
                        /* eslint-disable no-console */
                        console.error('Error: ', err);
                        /* eslint-enable no-console */
                    }
                }
            );

            this.subscriptions.push(s);
        }
    },

    notify() {
        this.updates.next();
    },

    componentWillUnmount() {
        this.propsStream.complete();
        this.updates.complete();
        this.subscriptions.forEach(sub => sub.unsubscribe());
        this.subscriptions.length = 0;
    },

    render() {},
});

const createComponentConstructor = Parent => {
    function ChildComponent(el) {
        Parent.call(this, el);
    }
    ChildComponent.prototype = Object.create(Parent.prototype, {
        constructor: { value: ChildComponent },
    });
    return ChildComponent;
};

export const baseCreateComponent = (transform, templateFnOrObj) => {
    const template = templateFnOrObj.render
        ? props => templateFnOrObj.render(props)
        : templateFnOrObj;
    const ChildComponent = createComponentConstructor(Component);
    ChildComponent.prototype.displayName =
        transform.name || template.name || template.displayName || 'Unknown';
    ChildComponent.prototype.render = function() {
        return template(this.state);
    };
    ChildComponent.prototype.getTransform = api => combine(transform(api));
    return ChildComponent;
};

export const createComponent = (...args) => {
    if (args.length >= baseCreateComponent.length) {
        return baseCreateComponent.apply(null, args);
    }
    return (...args2) => createComponent.apply(null, args.concat(args2));
};
