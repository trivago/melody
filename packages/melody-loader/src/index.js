/**
 * Copyright 2017 trivago N.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS-IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { toString, compile } from 'melody-compiler';
import { extension as CoreExtension } from 'melody-extension-core';
import { getOptions } from 'loader-utils';
import { isString, isObject } from 'lodash';

module.exports = function loader(content) {
    this.cacheable();

    const loaderOptions = getOptions(this) || {
        plugins: [],
    };

    const args = [this.resourcePath, content, CoreExtension];
    if (loaderOptions.plugins) {
        for (const pluginName of loaderOptions.plugins) {
            if (isString(pluginName)) {
                try {
                    args.push(require('melody-plugin-' + pluginName));
                } catch (e) {
                    this.emitWarning(
                        'Could not find plugin ' +
                            pluginName +
                            '. Expected name to be melody-plugin-' +
                            pluginName
                    );
                }
            } else if (isObject(pluginName)) {
                args.push(pluginName);
            } else {
                this.emitWarning(
                    'Value passed as Melody plugin must be string or object. ' +
                        pluginName +
                        ' of type ' +
                        typeof pluginName +
                        ' was given'
                );
            }
        }
    }

    try {
        const result = toString(compile.apply(null, args), content);
        return result.code;
    } catch (e) {
        this.emitError(e);
        return (
            'import {text} from "melody-idom"; export default { render(options) { text("Could not load ' +
            this.resourcePath +
            '"); console.error("Could not load ' +
            this.resourcePath +
            '", ' +
            JSON.stringify(e.message) +
            '); } };'
        );
    }
};
