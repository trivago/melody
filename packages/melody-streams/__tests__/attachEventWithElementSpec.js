/**
 * Copyright 2019 trivago N.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS-IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { attachEventWithElement } from '../src';
import { applyGradualyAndComplete } from './util/testHelpers';
import { createMouseEvent } from './util/mouseEvent';

const dispatchClick = createMouseEvent('click');
const dispatchMouseEnter = createMouseEvent('mouseenter');

describe('attachEventWithElement', () => {
    it('should attach a click handler', async () => {
        const el = document.createElement('div');
        const [refHandler, subj, originEl] = attachEventWithElement('click');
        refHandler(el);
        applyGradualyAndComplete(subj, dispatchClick(el), [
            undefined,
            undefined,
            undefined,
        ]).then(handlers => {
            expect(JSON.stringify(handlers)).toMatchSnapshot();
            expect(el).toBe(originEl.value);
        });
    });

    it('should attach multiple handlers', async () => {
        const el = document.createElement('div');
        const [refHandler, subj, originEl] = attachEventWithElement(
            'click',
            'mouseenter'
        );
        refHandler(el);
        applyGradualyAndComplete(
            subj,
            [dispatchClick(el), dispatchMouseEnter(el)],
            [undefined, undefined, undefined]
        ).then(handlers => {
            expect(JSON.stringify(handlers)).toMatchSnapshot();
            expect(el).toBe(originEl.value);
        });
    });
});
