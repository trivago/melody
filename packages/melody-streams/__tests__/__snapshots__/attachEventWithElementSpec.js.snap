// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`attachEventWithElement should attach a click handler 1`] = `"[{\\"type\\":\\"click\\"},{\\"type\\":\\"click\\"},{\\"type\\":\\"click\\"}]"`;

exports[`attachEventWithElement should attach multiple handlers 1`] = `"[{\\"type\\":\\"click\\"},{\\"type\\":\\"click\\"},{\\"type\\":\\"click\\"},{\\"type\\":\\"mouseenter\\"},{\\"type\\":\\"mouseenter\\"},{\\"type\\":\\"mouseenter\\"}]"`;
